be/src/storage/lake/lake_persistent_index.cpp

```diff
@@ -0,0 +1,294 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "storage/lake/lake_persistent_index.h"
+
+#include "gen_cpp/persistent_index.pb.h"
+#include "storage/chunk_helper.h"
+#include "storage/lake/meta_file.h"
+#include "storage/primary_key_encoder.h"
+#include "storage/tablet_meta_manager.h"
+
+namespace starrocks {
+namespace lake {
+
+Status LakePersistentIndex::load_from_lake_tablet(starrocks::lake::Tablet* tablet, const TabletMetadata& metadata,
+                                                  int64_t base_version, const MetaFileBuilder* builder) {
+    MonotonicStopWatch timer;
+    timer.start();
+
+    if (!is_primary_key(metadata)) {
+        LOG(WARNING) << "tablet: " << tablet->id() << " is not primary key tablet";
+        return Status::NotSupported("Only PrimaryKey table is supported to use persistent index");
+    }
+
+    PersistentIndexMetaPB index_meta;
+
+    // persistent_index_dir has been checked
+    Status status = TabletMetaManager::get_persistent_index_meta(
+            StorageEngine::instance()->get_persistent_index_store(), tablet->id(), &index_meta);
+    if (!status.ok() && !status.is_not_found()) {
+        return Status::InternalError("get tablet persistent index meta failed");
+    }
+
+    // There are three conditions
+    // First is we do not find PersistentIndexMetaPB in TabletMeta, it maybe the first time to
+    // enable persistent index
+    // Second is we find PersistentIndexMetaPB in TabletMeta, but it's version is behind applied_version
+    // in TabletMeta. It could be happened as below:
+    //    1. Enable persistent index and apply rowset, applied_version is 1-0
+    //    2. Restart be and disable persistent index, applied_version is update to 2-0
+    //    3. Restart be and enable persistent index
+    // In this case, we don't have all rowset data in persistent index files, so we also need to rebuild it
+    // The last is we find PersistentIndexMetaPB and it's version is equal to latest applied version. In this case,
+    // we can load from index file directly
+
+    if (status.ok()) {
+        // all applied rowsets has save in existing persistent index meta
+        // so we can load persistent index according to PersistentIndexMetaPB
+        EditVersion version = index_meta.version();
+
+        // Compaction of Lake tablet also generate a new version
+        // here just use version.major so that PersistentIndexMetaPB need't to be modified,
+        // the minor is meaningless
+        if (version.major() == base_version) {
+            // If format version is not equal to PERSISTENT_INDEX_VERSION_2, this maybe upgrade from
+            // PERSISTENT_INDEX_VERSION_2.
```

Comment:
upgrade from PERSISTENT_INDEX_VERSION_1?

Reply:
This is the version of the persistent index format, not the index type, which is consistent with SHARD NOTHING. The format version of SHARD NOTHING is PERSISTENT_INDEX_VERSION_2